cirq/sim/wave_function.py

```diff
@@ -282,3 +286,40 @@ def _validate_indices(num_qubits: int, indices: List[int]) -> None:
     if any(index >= num_qubits for index in indices):
         raise IndexError('Out of range indices, must be less than number of '
                          'qubits but was {}'.format(indices))
+
+
+def sample_terminal_measurements(
+        circuit: circuits.Circuit,
+        last_step_result: 'simulator.StepResult',
+        repetitions: int) -> Dict[str, List]:
+    """Sample from measurements in the given circuit.
+
+    This should only be called if the circuit has only terminal measurements.
+
+    Args:
+        circuit: The circuit to sample from.
+        last_step_result: The XmonStepResult from which to sample. This should
+            be the step at the end of the circuit. Can be None if no steps were
+            taken.
+        repetitions: The number of time to sample.
+
+    Returns:
+        A dictionary from the measurement keys to the measurement results.
+        These results are lists of lists, with the outer list corresponding to
+        the repetition, and the inner list corresponding to the qubits as
+        ordered in the measurement gate.
+    """
+    if last_step_result is None:
```

Comment:
If the user calls this with a non-terminal measurement circuit, it silently gives bad results?

Reply:
Moved into step result.  The semantics are that you need to pass it a set of measurement results that act on disjoint qubits.  Since it is on a step result it s clear that it means (at this time).  The part about being terminal is handled by the simulator class.